openedx/core/djangoapps/schedules/resolvers.py

```diff
@@ -1,136 +1,313 @@
+from itertools import groupby
+import logging
+
+import attr
+from django.conf import settings
+from django.contrib.auth.models import User
+from django.contrib.staticfiles.templatetags.staticfiles import static
+from django.core.urlresolvers import reverse
+from django.db.models import F, Q
+from django.utils.formats import dateformat, get_format
+
 
 from edx_ace.recipient_resolver import RecipientResolver
-from edx_ace.utils.date import serialize
-
-from openedx.core.djangoapps.schedules.models import ScheduleConfig
-from openedx.core.djangoapps.schedules.tasks import (
-    DEFAULT_NUM_BINS,
-    RECURRING_NUDGE_NUM_BINS,
-    UPGRADE_REMINDER_NUM_BINS,
-    COURSE_UPDATE_NUM_BINS,
-    recurring_nudge_schedule_bin,
-    upgrade_reminder_schedule_bin,
-    course_update_schedule_bin,
-)
+from edx_ace.recipient import Recipient
+
+from courseware.date_summary import verified_upgrade_deadline_link, verified_upgrade_link_is_valid
+from openedx.core.djangoapps.monitoring_utils import function_trace, set_custom_metric
+from openedx.core.djangoapps.schedules.config import COURSE_UPDATE_WAFFLE_FLAG
+from openedx.core.djangoapps.schedules.exceptions import CourseUpdateDoesNotExist
+from openedx.core.djangoapps.schedules.models import Schedule
 from openedx.core.djangoapps.schedules.utils import PrefixedDebugLoggerMixin
-from openedx.core.djangoapps.site_configuration.models import SiteConfiguration
+from openedx.core.djangoapps.schedules.template_context import (
+    absolute_url,
+    get_base_template_context
+)
+
+from request_cache.middleware import request_cached
+from xmodule.modulestore.django import modulestore
+
+
+LOG = logging.getLogger(__name__)
 
+DEFAULT_NUM_BINS = 24
+RECURRING_NUDGE_NUM_BINS = DEFAULT_NUM_BINS
+UPGRADE_REMINDER_NUM_BINS = DEFAULT_NUM_BINS
+COURSE_UPDATE_NUM_BINS = DEFAULT_NUM_BINS
 
+
+@attr.s
 class BinnedSchedulesBaseResolver(PrefixedDebugLoggerMixin, RecipientResolver):
     """
     Starts num_bins number of async tasks, each of which sends emails to an equal group of learners.
 
     Arguments:
         site -- Site object that filtered Schedules will be a part of
         current_date -- datetime that will be used (with time zeroed-out) as the current date in the queries
+        async_send_task -- celery task function which this resolver will call out to
 
     Static attributes:
-        async_send_task -- celery task function which this resolver will call out to
         num_bins -- the int number of bins to split the users into
         enqueue_config_var -- the string field name of the config variable on ScheduleConfig to check before enqueuing
     """
-    async_send_task = None  # define in subclass
-    num_bins = DEFAULT_NUM_BINS
-    enqueue_config_var = None  # define in subclass
+    async_send_task = attr.ib()
+    site = attr.ib()
+    target_datetime = attr.ib()
+    day_offset = attr.ib()
+    bin_num = attr.ib()
+    org_list = attr.ib()
+    exclude_orgs = attr.ib(default=False)
```

Comment:
Consider renaming this to `should_exclude_orgs` so it's more clear that this is a boolean value and not a list of orgs.

Reply:
That's all going to go away in my next PR anyway, so I'm going to just leave it for now.